crates/trigger-http/src/server.rs

```diff
@@ -138,12 +148,46 @@ impl<F: RuntimeFactors> HttpServer<F> {
 
     /// Serve incoming requests over the provided [`TcpListener`].
     pub async fn serve(self: Arc<Self>) -> anyhow::Result<()> {
-        let listener = TcpListener::bind(self.listen_addr).await.with_context(|| {
-            format!(
-                "Unable to listen on {listen_addr}",
-                listen_addr = self.listen_addr
-            )
-        })?;
+        let listener: TcpListener = match TcpListener::bind(self.listen_addr).await {
+            Ok(listener) => listener,
+            Err(err) => {
+                if self.find_free_port && err.kind() == ErrorKind::AddrInUse {
+                    let mut found_listener = None;
+                    let mut addr = self.listen_addr;
+                    for _ in 1..=9 {
+                        if addr.port() == u16::MAX {
+                            anyhow::bail!(
+                                "Couldn't find a free port as we've reached the maximum port number. Consider retrying with a lower base port."
+                            );
+                        }
+
+                        addr.set_port(addr.port() + 1);
+
+                        match TcpListener::bind(addr).await {
+                            Ok(listener) => {
+                                found_listener = Some(listener);
+                                break;
+                            }
+                            Err(err) => {
+                                if err.kind() == ErrorKind::AddrInUse {
+                                    continue;
+                                }
+                                anyhow::bail!("Unable to listen on {}", self.listen_addr);
+                            }
+                        }
+                    }
+
+                    found_listener.ok_or_else(|| anyhow::anyhow!(
+                        "Couldn't find a free port in the range {}-{}. Consider retrying with a different base port.",
+                        self.listen_addr.port(),
+                        self.listen_addr.port() + 9
```

Comment:
Duplicate magic number (with line 157).  Make this a const, or if you adopt the helper function, it could even be a parameter passed to the helper.